mlrPlayground blog post [WIP] (#11)
mlr-org/mlr-web: content/docs/2019-06-23-introducing-mlrplayground.Rmd

---
title: "Introducing mlrPlayground"
authors: ["sebastian-gruber"]

> date: "2019-06-23"

Reviewer: FYI: If you date the date back in time, you get a preview rendered by Netlify in the Github PR. 
This is nice for getting an impression on how your post will look like. 
Especially for reviewers who did not do a local rendering.

output:
  blogdown::html_page:
    toc: true
categories:
  - r-bloggers
tags:
  - shiny
  - visualization
  - mlr
  - teaching
  - classification
  - regression
  - hyperparameter
---

<img src="https://i.imgur.com/MPEwCJ2.gif" width=800 />
<br/>

## First of all
You may ask yourself how is this name ‘mlrPlayground’ even justified?

> What a madman puts two such opposite terms in a single word and expects people to take him seriously?

Reviewer: I like catchy starts but this one is maybe a bit over the top :) At least "madman" should be replaced.

Author: done

> I assume most of you know ‘mlr’, for those who don’t: it’s a framework for machine learning tasks offering a huge variety of different learners based on the best packages available in R, it’s about productivity, simplicity, and extendability.

Reviewer: its's -> It is

"a huge variety of different learners based on the best packages available in R" -> Let's just say it offers a lot of tools to simplify machine-learning tasks in R

"simplicity, and extendability." maybe we should leave the first word out for mlr ;). The second should be phrased differently.

> Quite the opposite from a place, where you can play with your best friends, make new friends, live out your fantasies and just have a great time the whole day until your parents pick you up.

Reviewer: "live out your fantasies" sounds a bit too sexual here :)

Author: If read out of context, I would agree, but certainly not between "make new friends" and "getting picked up by your parents".

Well, for most of the readers here this may not be the case anymore – we know, we are still young in our heart, but let’s be honest … 

> For sure, we all have those memories and definitely have certain associations by the word ‘Playground’.

Reviewer: ```suggestion
For sure, we all have those memories and definitely have certain associations with the word ‘Playground’.
```

So what is it then with this thing called ‘mlrPlayground’?
<br/>

## The idea

> The thought behind this project was to offer a platform in the form of a <a href="https://shiny.rstudio.com/">Shiny</a> web application, where a user can try out different kinds of learners provided by the mlr package on a small set of destinct and tunable regression and classification tasks, making it possible to observe the prediction/performance behaviour based on changes on the task, the learner or the learner's hyperparameters.

Reviewer: Split this sentence into three separate ones. 

"where" -> in which

> In the best case the user will gain new insights and deeper understanding of how each learner works, where it’s advantages are and also very importantly the cases when the learner fails miserably.

Reviewer: ```suggestion
The user is able to gain new insights and a deeper understanding of how a learner performs, where it’s advantages are and in which cases the learner might fail.
```

> There are a lot of different settings we want to offer in the user interface, and so – to not remove the fun of our playground – a huge effort went into making the UI as aesthetically pleasing and as smooth as possible.

Reviewer: ```suggestion
There are a lot of different settings we want to offer in the user interface, and so – to not remove the fun of our playground – a huge effort went into creating an aesthetically pleasing UI.
```

> To achieve this, a website template was downloaded from <a href="https://templated.co/">Templated</a> and used as a baseline for design. After extending the template with missing CSS classes, most of the used shiny widgets have been overwritten – or even completely replaced –, offering refreshingly new visuals for the well known shiny framework.

Reviewer: Paragraph break before this sentence.

Always start sentences on a new line (for review purposes).

```suggestion
To achieve this, a website template was downloaded from <a href="https://templated.co/">Templated</a> and used as the baseline design. 
After extending the template with missing CSS classes, most of the used shiny widgets have been overwritten – or even completely replaced –, offering refreshingly new visuals for the well-known shiny framework.
```

> For the smoothness part, an object oriented <a href="https://cran.r-project.org/web/packages/R6/index.html">R6</a> class system with reactive attributes was engineered for the backend to give a well-defined framework of which elements should trigger what evaluation; an otherwise extremely tiresome and error-prone task for dozens of different UI elements.

Reviewer: Split into at least two sentences.

```suggestion
For the smooth feel of the app, an object oriented <a href="https://cran.r-project.org/web/packages/R6/index.html">R6</a> class system with reactive attributes was engineered for the backend to give a well-defined framework of which elements should trigger what evaluation; an otherwise extremely tiresome and error-prone task for dozens of different UI elements.
```

After all ‘mlrPlayground’ may not be as fun as a real playground, but you are also not as likely to hurt yourself and it is definitely more entertaining than looking at boring pictures of learners in a book.
<br/>

## The features

> In the following an extended overview of almost all available features is presented with the help of animated pictures.

Reviewer: ```suggestion
In the following an extended overview of most available features is presented with the help of animated pictures.
```

###### Task selection

> When running the app and scrolling down you can basically see two big boxes with content: the left one is for the task and learner settings, the right one for an overview of the learner’s bevahiour.

Reviewer: - Spell check

- Always start with a capital letter after a colon

> For the first look it may be of help (even though not necessary) to explore the tasks.

Reviewer: This sounds very German, please rephrase :)

> After finding the right button (I leave this task to the motivated reader) a panel with all available datasets is extended over our screen.

Reviewer: I -> we

> Here, we can choose the tasktype of either classification with two covariables or regression with one covariable and several other parameters influencing the dataset: size, noise and the train/test split.

Reviewer: ```suggestion
Here, we can choose the task type: The user can select from of a variety of different task types which again contain further settings such as dataset size, noise ratio or the train/test split ratio.
```

Author: I'll keep my version as I don't like the repeated "task type"

> <img src="https://i.imgur.com/ZV1wlBE.gif" width=800 />

Reviewer: The GIF is great but the resolution could be better. Is it possible to create the GIF with a higher res?
While at it: It would be cool if there was no "windows watermark" in the bottom right corner.

<br/>

###### Learner selection

> Back at the main panel we can then select a learner for our task – due to limited manpower not all available in mlr have made it into this selection, expect extensions in the future. Now, it may take a few seconds because several things are happening in the background:

Reviewer: ```suggestion
Back at the main panel we can then select a learner for our task – due to limited manpower not all available learners in _ml_r made it into this selection.
You can expect additions in the future. 
Now, the app is ready for processing:
```

<ul>
<li><p>a model is trained on the training set</p></li>
<li><p>different kinds of performance measures are calculated on the test set</p></li>

> <li><p>a grid of points is used for inference to create a plot showing the prediction surface.</p></li>

Reviewer: I assume this sentence might confuse the reader. Can it be rephrased in more easy words?

Also is there a reason why you are using HTML lists here?

</ul>

<img src="https://i.imgur.com/7fAdJEu.gif" width=800 />
<br/>

###### Performance measures

> The first visual change after calculations have finished is the appearance of the prediction plot, but this is not the only part that's getting updated. Besides the plots are a whole bunch of different performance measures to explore by clicking on the equally named bar on the right hand side. These give the opportunity to manually tune each learner on the tasks or compare how each measure behaves on certain scenarios. Not only will you be able to understand the learner better, but also each measure. What's particularly interesting for the latter is the ratio of the train to the test data settable in the task selection window.

Reviewer: ```suggestion
The first visual change after calculations have finished is the appearance of the prediction plot. 
Besides there are now a whole bunch of different performance measures to explore by clicking on the equally named bar on the right hand side. 
These provide the opportunity to manually tune each learner on the tasks or compare how each measure behaves in certain scenarios. 
You will not only be able to understand the learner better, but also each measure. 
What's particularly interesting for the latter is the configurable ratio of train and test data in the task selection window.
```

<img src="https://i.imgur.com/WGqQsqt.gif" width=800 />
<br/>

###### Hyperparameters

> If you want to see the effect of hyperparameters on the plot/performance measures, simply click on the small cog next to the learner selection. Now, the left panel is replaced by a window with all available hyperparameters – changes of these have an immediate effect of recalculating the whole model and updating all plots/numbers.

Reviewer: ```suggestion
If you want to see the effect of hyperparameters on the plot/performance measures, simply click on the small cog next to the learner selection. 
Now, the left panel is replaced by a window with all available hyperparameters – changes here will have an immediate effect: The whole model will be rerun and all plots/numbers will get updated.
```

<img src="https://i.imgur.com/YvODxIt.gif" width=800 />

###### Second learner

> Only observing changes of a single learner may become a bit boring after time, so a small button is available to easily double the fun: Pressing “Add learner” gives the choice of a second learner with a second plot, making it possible to compare the prediction surfaces of two different learners on the same task right next of each other. Incredible. A bit.

Reviewer: ```suggestion
Only observing changes of a single learner may become a bit boring after time, so an option is available to easily double the fun: Pressing “Add learner” gives the choice of adding a second learner with a second plot, making it possible to compare the prediction surfaces of two different learners on the same task right next of each other. 
Exciting!
```

<img src="https://i.imgur.com/Kho9pDh.gif" width=800 />
<br/>

But wait, there’s still more to offer!

###### Learning curve and ROC

> The curious user can change the shown plots on the right panel by switching to the tabs “Learning Curve” (classification and regression) or “ROC” (only classification). The former tells you how well the learner does, or the learners do compared to each other, with respect to user selected performance measurements (y-axis) on a fraction of the original train data (x-axis). Every additionally selected measure gives an additional plot. Not groundbreaking, but handy.

Reviewer: ```suggestion
The user can change the created plots on the right panel by switching to the tabs “Learning Curve” (classification and regression) or “ROC” (only classification). 
The former tells you how well the learner performed or how multiple learners compare to each other.
All of this happens with respect to the user selected performance measures (y-axis) on a fraction of the original train data (x-axis). 
Every additionally selected measure results in an additional plot. 
```

measures != measurements

> Last but not least, the ROC curve is only an ROC curve with its default settings for the x- and y-axis: Plotting the true positive rate against the false positive rate, giving good advice of how well the learner can separate the classes (the bigger the area under the curve, the better).

Reviewer: ```suggestion
Last but not least, the ROC curve is only an ROC curve with its default settings for the x- and y-axis: Plotting the true positive rate against the false positive rate.
It shows how well the learner can separate the classes (the bigger the area under the curve, the better).
```

> Change the settings and you’ll get quite exotic variations, whose names are beyond the scope of the author.

Reviewer: Just omit that :)

<img src="https://i.imgur.com/qs8E51L.gif" width=800 />
<br/>

## Usage

If you want to explore the mlrPlayground on your own, you have two possibilities:

Either visit the website https://compstat-lmu.shinyapps.io/mlrPlayground/ (big thanks to the department of computational statistics at the LMU for hosting)

or

> install it locally on your own machine by ``devtools::install_github("SebGruber1996/mlrPlayground")``

Reviewer: ```suggestion
install it locally on your own machine by `remotes::install_github("SebGruber1996/mlrPlayground")`
```

devtools::install_github is deprecated.

and start with ``mlrPlayground::start()``.

> The first method is quicker, but may crash if too many people are using it at once, while the latter requires a long time for installation.

Reviewer: ```suggestion
The first method is quicker, but may crash if too many people access it at the same time, while the latter requires a local installation.
It's your choice. 
```

<br/>
<br/>

Thanks for spending your time reading this blog post instead of being on a real playground!
<br/> 
<br/>
<br/>

Side note: The app is still far from perfect.

> For example the UI is -- especially for the hyperparameters -- broken for small monitor resolutions. 

Reviewer: ```suggestion
The UI is -- especially for the hyperparameters -- not working well for small displays. 
```

"small monitor resolutions" do not exist. Either "small displays" or "low resolutions",

> Also some hyperparameter settings may still lead to a crash.

Reviewer: ```suggestion
Also some hyperparameter settings may cause the app to crash.
```

> Even if these things would be worked out, there are still a lot of learners missing in the app that are available in the mlr package.

Reviewer: ```suggestion
Even if these things would be solved, there are still a lot of learners missing in the app which are available in the _mlr_ package.
```

> If you feel motivated, I'll gladly accept every solid pull request proposed to https://github.com/SebGruber1996/mlrPlayground.

Reviewer: ```suggestion
Pull requests are welcome at https://github.com/SebGruber1996/mlrPlayground.
```
